pygfx/objects/_events.py

```diff
@@ -0,0 +1,314 @@
+from collections import defaultdict
+from enum import Enum
+from time import perf_counter
+from typing import Union
+
+
+class EventType(str, Enum):
+    # Keyboard
+    KEY_DOWN = "key_down"
+    KEY_UP = "key_up"
+    # Pointer
+    POINTER_DOWN = "pointer_down"
+    POINTER_MOVE = "pointer_move"
+    POINTER_UP = "pointer_up"
+    DOUBLE_CLICK = "double_click"
+    # Wheel
+    WHEEL = "wheel"
+    # Window
+    CLOSE = "close"
+    RESIZE = "resize"
+
+
+class Event:
+    """Event base class for creating events.
+
+    If a target is set, an event can bubble up through a hierarchy
+    of targets, connected through a ``parent`` property.
+    To prevent an event from bubbling up, use ``stop_propagation``.
+
+    Event implements the ``__getitem__` method which makes an event
+    behave like a (read-only) dictionary. This makes it a bit more
+    compatible with the jupyter_rfb event spec.
+
+    Custom events can have any fields. Any ``kwargs`` will be captured
+    and can be later retrieved with the square bracket notation, using
+    the same keyword.
+    """
+
+    def __init__(
+        self,
+        type: Union[str, EventType],
+        *,
+        bubbles=True,
+        target: "EventTarget" = None,
+        **kwargs,
+    ):
+        self._type = type
+        self._time_stamp = perf_counter()
+        self._bubbles = bubbles
+        self._target = target
+        self._current_target = target
+        self._cancelled = False
+        # Save extra kwargs to be able to look
+        # them up later with `__getitem__`
+        self._data = kwargs
+
+    @property
+    def type(self) -> str:
+        """A string representing the name of the event."""
+        return self._type
+
+    @property
+    def time_stamp(self) -> float:
+        """The time at which the event was created (in seconds). Not actual timestamp
+        so please only use for relative time measurements."""
+        return self._time_stamp
+
+    @property
+    def bubbles(self) -> bool:
+        """A boolean value indicating whether or not the event bubbles up through
+        the scene tree."""
+        return self._bubbles
+
+    @property
+    def target(self) -> "EventTarget":
+        """The object onto which the event was dispatched."""
+        return self._target
+
+    @property
+    def current_target(self) -> "EventTarget":
+        """The object that is currently handling the event. During event bubbling
+        this property will be updated whenever an event bubbles up in the hierarchy."""
+        return self._current_target
+
+    @property
+    def is_cancelled(self) -> bool:
```

Comment:
personally, I would be ok leaving off `is_` just so it's shorter, but it's your call :)

Reply:
It is also asymmetrical with `bubbles`, so yeah: I've removed the `is_`.